Automattic/woocommerce-services
# Install and activate a list of plugins during migrations

## Description
> When the Update button is clicked, it should initiate the background installer to start installing WC Shipping and then update the button state to `Updating` until the installation is complete.

This PR installs and activates  "WooCommerce Shipping" and WooCommerce Tax" upon clicking the "Update" button. It also deactivates WCS&T afterwards.

The order is:
1. Install WooCommerce Shipping.
2. Install WooCommerce Tax.
3. Activates WooCommerce Shipping.
4. Activates WooCommerce Tax.
5. Deactivate WCS&T

## TODO
- Upon clicking "Update", it should update a flag to indicate that the user is migrated. This will be done via another PR. 
- Error handling if install/activation fails. Where do the errors go?

### Technical considerations
I contemplated using `@wordpress/apiFetch` or not. I tried using it at first thinking that:
1. It fits better with the wp-data way.
6. It handles nonces.
Unfortunately, WCS&T does not include the `@wordpress/apiFetch` package. I tried installing this (first commit) and can verify that it also doesn't automatically handle nonce. 

At the end, I removed the package and opted to use the native `fetch` instead. In this repo, we have a library that talks to the connect server. This client uses a nonce. This PR utilizes the same nonce. 

### Related issue(s)
N/A

### Testing activation
1. Go to `woocommerce-services/classes/class-wc-connect-service-settings-store.php` and set the following to `true`
```
		public function is_eligible_for_migration() {
			return true;
		}
```
2. Make sure you have "WooCommerce Shipping" and "WooCommerce Tax" in your plugins folder. We can't test the "install" part because they aren't part of dotorg yet.
3. Go to an individual order where you can see the WCS&T shipping banner.
4. Click "Create shipping label" 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/38dcd472-8723-4692-bfb1-da77e896e2da)
5. Click "Update now"
![image](https://github.com/Automattic/woocommerce-services/assets/572862/327a9162-e43f-4f26-be6f-6462b56f8b75)
6. You should see 3 requests go by the network tab. `install`, `activate` and a `woocommerce-service`
![image](https://github.com/Automattic/woocommerce-services/assets/572862/b1553d89-bb58-4f9b-a3b9-2d14ef31eaf8)
7. Confirm that you are redirected to the plugins page and confirm that WCS&T is deactivated, WooCommerce Shipping and WooCommerce tax are activated. 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/2ff5d3ae-2359-41bd-b523-92e50918c387)

### Testing installation
To test installation, we can test it against the `hello-dolly` plugin. 
1. Go to http://localhost/wp-admin/plugins.php and delete your "Hello Dolly" plugin.
2. Open `client/components/migration/feature-announcement.jsx`, change line 34 from `const plugins = 'woocommerce-shipping,woocommerce-tax';` to `const plugins = 'hello-dolly';`. This will now install and activate hello dolly.
3. Repeat step 3-7 above and confirm "Hello Dolly" is installed and activated. 

### Checklist
